initrd/etc/ash_functions

```diff
@@ -53,6 +53,154 @@ preserve_rom() {
+			echo
+			#TODO: change all passphrase prompts in codebase to include -r to prevent backslash escapes
+			read -r -s -p "Please enter GPG Admin PIN needed to use the GPG backup thumb drive: " gpg_admin_pin
+			echo
+			#prompt user to select the proper encrypted partition, which should the first one on next prompt
+			warn "Please select encrypted LUKS on GPG key material backup thumb drive (not public labeled one)"
+			mount-usb --pass "$gpg_admin_pin" || die "Unable to mount USB with GPG Admin PIN"
+			echo "++++ Testing detach-sign operation and verifiying against fused public key in ROM"
+			gpg --pinentry-mode=loopback --passphrase-file <(echo -n "${gpg_admin_pin}") --import /media/subkeys.sec >/dev/null 2>&1 ||
+				die "Unable to import GPG private subkeys"
+			#Do a detach signature to ensure gpg material is usable and cache passphrase to sign /boot from caller functions
+			dd if=/dev/urandom of="$CR_NONCE" bs=20 count=1 >/dev/null 2>&1 ||
+				die "Unable to create $CR_NONCE to be detach-signed with GPG private signing subkey"
+			gpg --pinentry-mode=loopback --passphrase-file <(echo -n "${gpg_admin_pin}") --detach-sign "$CR_NONCE" >/dev/null 2>&1 ||
+				die "Unable to detach-sign $CR_NONCE with GPG private signing subkey using GPG Admin PIN"
+			#verify detached signature against public key in rom
+			gpg --verify "$CR_SIG" "$CR_NONCE" > /dev/null 2>&1 && \
+                                echo "++++ Imported private subkeys match public key fused in rom and can be used under Heads"  || \
+			 	die "Unable to verify $CR_SIG detached signature against public key in ROM"
+			#Wipe any previous CR_NONCE and CR_SIG
+			shred -n 10 -z -u "$CR_NONCE" "$CR_SIG" >/dev/null 2>&1 || true
+			#TODO: maybe just an export instead of setting /etc/user.config otherwise could be flashed in weird corner case situation
+			set_user_config "CONFIG_GPG_KEY_BACKUP_IN_USE" "y"
+			umount /media || die "Unable to unmount USB"
+			return
+		fi
+	fi
+
+	# setup the USB so we can reach the GPG card
+	enable_usb
+
+	echo -e "\nVerifying presence of GPG card...\n"
+	# ensure we don't exit without retrying
+	errexit=$(set -o | grep errexit | awk '{print $2}')
+	set +e
+	gpg --card-status >/dev/null
+	if [ $? -ne 0 ]; then
+		# prompt for reinsertion and try a second time
+		read -n1 -r -p \
+			"Can't access GPG key; remove and reinsert, then press Enter to retry. " \
+			ignored
+		# restore prev errexit state
+		if [ "$errexit" = "on" ]; then
+			set -e
+		fi
+		# retry card status
+		gpg --card-status >/dev/null ||
+			die "gpg card read failed"
+	fi
+	# restore prev errexit state
+	if [ "$errexit" = "on" ]; then
+		set -e
+	fi
+}
+
+gpg_auth() {
+	if [[ "$CONFIG_HAVE_GPG_KEY_BACKUP" == "y" ]]; then
+		TRACE "Under /etc/ash_functions:gpg_auth"
+		# If we have a GPG key backup, we can use it to authenticate even if the card is lost
+		echo >&2 "!!!!! Please authenticate with OpenPGP card/backup media to prove you are the owner of this machine !!!!!"
```

Comment:
@tlaurion It might be easiest to discuss this feature if we move it to a separate PR dependent on this one.  These are good ideas but there's already a lot here with the GPG keygen improvements :wink: 

I think it is unexpected to tie this feature to the choice of creating a GPG key backup during OEM reset - it's surprising, has a bit of lockout risk, and it also means OEM reset is the only way to activate/deactivate this feature.

Could it be a normal configuration menu option?

Regarding the feature behavior, I think these are good ideas but it is worth agreeing on specific security goals.  There is also overlap with restricted boot, so we should ensure those work well together (maybe changing RB to fit in well with this feature).

Is the goal that an attacker can't observe the firmware or encrypted disk contents without creating evidence of tampering?  If so, you'd also need to prevent flashing an upgrade, maybe some configuration options, and possibly other things, like RB does.  (I think there are enough "other things" that this would probably be best served as its own PR.)

Reply:
> Could it be a normal configuration menu option?


I think we should add a CONFIG_GPG_AUTH=y and that both CONFIG_GPG_AUTH= and CONFIG_HAVE_GPG_KEY_BACKUP=y should be checked by gpg_auth to enforce authentication in code? 

I think CONFIG_GPG_AUTH=y could be a configuration option toggle if CONFIG_HAVE_GPG_KEY_BACKUP=y is already present (and confirms user has a backup to use in case USB Security dongle is lost.

The whole idea of gpg auth is to be offered only if GPG key material backup is available. I may have took a shorcut by using the same provisioning variable for gpg_auth enforcement, but I think the PR should offer gpg_auth out of the box.

@JonathonHall-Purism ?

Reply:
:eyes: This is getting really long already and I have not gotten through the use cases you listed yet, posting this to break it up as I continue to go through them.  May edit this if some of my thoughts are already answered :thinking: 

>I think we should add a CONFIG_GPG_AUTH=y and that both CONFIG_GPG_AUTH= and CONFIG_HAVE_GPG_KEY_BACKUP=y should be checked by gpg_auth to enforce authentication in code?

Agree that it should be a separate config but IMO "GPG_AUTH" is too vague, and it's also a "how".  Other things, like the content of /boot, are also authenticated with GPG or could be.  Since the goal of this feature is to prevent tampering by requiring authentication before any action that could tamper, how about CONFIG_TAMPER_PREVENTION?

I'm not sure it needs to be tied to CONFIG_HAVE_GPG_KEY_BACKUP, but probably better safe than sorry, we can leave it tied to that initially and expand later if it makes sense.

>I think the PR should offer gpg_auth out of the box.

Why does that matter?  If you just don't want to merge the keygen improvements without GPG auth for some reason, it could still be two PRs :thinking: 

That said it's not critical either, but this is turning into a monster thread already buried inside of another feature's PR :thinking: 

>Security goals: have authentication enforced prior of doing stuff that are not necessarily leaving tampering traces without restricting freedoms, and where left feasible actions can not wipe data (DOS) unless user authenticates himself, leaving physicla tampering/theft as the only possible threats to daily usage and leaving laptop unattended in a reasonably secure location.

This is great :star_struck:   So we want to prevent a physically-present attacker from disabling the system (in the absence of physical tampering, which as always requires physical countermeasures).  That's different from RB as well.  Booting the system authenticates by unlocking the disk (either normally or via TPM LUKS unlock), and we assume that the OS will not allow taking any destructive actions until the disk is unlocked (e.g. if it drops to a recovery shell after N failed attempts to unlock /, that'd be a problem).

So I think we need to prevent the following.  We could gate them on authentication or require that the feature is disabled to access them:
* External booting :heavy_check_mark: 
* Recovery shell :heavy_check_mark: 
* Flashing firmware :point_left:  Otherwise you could "upgrade" to a tampered ROM defeating this
* OEM reset
* Changing config settings - at least some of them, definitely this feature itself
* Boot menu (can't allow OS recovery options as you'd get an initramfs shell)
* Booting when no default is set, since it offers the boot menu (might be covered by boot menu)
* Reset TPM/TOTP/HOTP
* Maybe GPG menu?
* Anything else?

There is a lot of overlap with RB, and I can imagine probably future features will refine this policy even more.  I wonder if we could structure it around some sort of security policy, where operations indicate what functionality is trying to be accessed, and we evaluate a policy to see if that is allowed based on the current config and authentication.

E.g. some operations could be:

| Operation | Description | Normal | Restricted boot | Tamper prevention |
|---|---|---|---|---|
| Boot default | We assume this has some form of protection already before granting access | Allowed | Allowed | Allowed |
| Configure policy | E.g. change RB / TP / other relevant settings | Allowed | Prevented, except disabling RB | Requires auth (?) |
| Boot menu | Recovery shell boot could tamper | Allowed | Allowed | Prevented (?) |
| Tamper | Totally unrestricted access like flashing firmware / recovery shell | Allowed | Prevented | Requires auth |
| ... | Others? | ? | ? | ? |

So, e.g. if you are about to show the boot menu, call `check_policy boot_menu`, and `check_policy` would evaluate the configuration of RB / TP, authenticate if needed, etc.

Otherwise I think embedding the business logic for each of these into the scripts actually doing the work could quickly get out of hand, and we would end up with a lot of holes.

Reply:
>Prior of Restricted Boot mode, the goal of authenticated Heads was to enforce authentication prior of flashing, recovery shell and booting anything else the user has not directly enforced policies for. So as current state of this PR, this is acessing recovery shell and booting from USB: tails, Archlinux and other ditro signed isos under Heads could leave the unnatended user's computer useable for things the user would not want his computer to be used for without explicitely permitting it himself prior of doing so.

OK, so maybe this combines with what I thought above - you envision this being used in conjunction with Restricted Boot?  I had thought of it as another alternative, but maybe this makes more sense.

>where initramfs can still land in really minimalist in-ram OS, where limited expected damage can be done

You could easily wipe the disk from the initramfs shell.  But I think it's also reasonable to punt this to the OS for an initial implementation of the feature and consider this in another step.

>without physically tampering the device or without flashing another firmware without preserving settings or wiping configuration options, all of which requires flashing and tampers measurements and prevent correct unsealing of operations on next boot.

I think flashing even preserving settings needs to be prevented.  Flashing erasing settings can probably be allowed since it would erase the GPG keyring, which would be hard to reconstruct.

If we allow flash preserving settings, you could flash a firmware that ignores the tampering settings, use it to tamper, then flash back keeping the settings.  Admittedly you'd have to know what the original firmware was (but it does tell you the version), and you'd rely on being able to put the CBFS files back in the same place, but it's pretty predictable I think.

>As for proper integration of gpg_auth under Restricted Boot(RB), I think this can be done in iterative PR

OK yes, agree, we can iterate to get to the end goal.  So we don't have to check all the boxes I listed (and potentially others) initially.

Reply:
#1520 has discussion for the thoughts above, and we can continue that even if the PR is merged.  The current state of this PR is a good initial implementation, so resolving this :+1:

Reply:
Put those discussion points under https://github.com/linuxboot/heads/discussions/1520